### ISSUE TITLE
Request: create vmware_guest with defined powerstate powerdon|powerdoff

### ISSUE THREAD

**Issue by harvey637** [ansible/ansible#54045]
##### SUMMARY
create a new system on VMware with "vmware_guest" and keep it in the state "poweredoff"

##### ISSUE TYPE
Feature request: make deployment stop after creation of vmware_guest in poweredoff mode

##### COMPONENT NAME
module: vmware_guest

##### ADDITIONAL INFORMATION
Relating to the issue #50923 it is still not possible to create vmware systems with several disks on different datastores (datastore_autoselect allways uses the first datastore, see #50923).

This needs the creation of a system to be suspended after creation of the system with WRONG disk, delete the wrong disk and manually create the correct disk.
Only after this is done the system can be poweredon and automatical installation (e.g. kickstart)
is allowed to start.

At present this can only be accomplished with ctrl-c breaking the playbook and resetting the system in the vmware_gui.

The state "poweredon|powerdoff|present|..." is used in a strange way:
Setting it to "powerdoff" while CREATING a new vmware guest is simply ignored. But the same state: poweredoff an a ALREADY CREATED vmware_guest really switch off the guest NOW.

So this state can only be used with a existing guest, it is of no use (or even contrary to my whish)
to a to be created vmware guest:

Proposal: Add a flag (e.g. creation_state: [powerdon|poweredoff] default('powerdon')) to the CREATING of a vmware_guest, how the state after creation
of a new guest should be. IGNORE this flag on a INSTALLED guest.

And/Or make creation of guest possible with two or more different datastores.
Also please explain, which flags only work while creating, which work while updating.
(Updating a "created at date_time string" in annotation makes no sense in subsequent playbook runs)
cu
Harvey

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/vmware/vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @Tomorrow9 @ckotte @dav1x @lparkes @nerzhul @pdellaert @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@harvey637, just so you are aware we have a dedicated Working Group for vmware.
You can find other people interested in this in `#ansible-vmware` on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Akasurde** [comment]
@harvey637 Thanks for reporting this feature. Did you try the #50923 ? Let me know.

needs_info

**Comment by harvey637** [comment]
@Akasurde 
yes, see my answer in #50923.
In a way both belong togather. As I was NOT able to autoprovision a vmware guest with vmware_guest.py with two disks on different datastore (large harddisk and separate ssd was the
reason for this) I needed to stop the installation AFTER provisioning(creating) new vm_guest and 
BEFORE kickstarting the OS-installation.
Identified the problem, as only "self.params('disk'][0][datastore]" is used to create ALL disks, even if they have there own datastore/autoselect_datastore
cu
Harvey

**Comment by ansibot** [comment]
cc @goneri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @pgbidkar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ngp-star** [comment]
hi @harvey637 
can  you just explain how to reproduce the issue , and can you just explain the following scenario what you mentioned in additional information section of this issue.

Relating to the issue #50923 it is still not possible to create vmware systems with several disks on different datastores (datastore_autoselect allways uses the first datastore, see #50923).

This needs the creation of a system to be suspended after creation of the system with WRONG disk, delete the wrong disk and manually create the correct disk.
Only after this is done the system can be poweredon and automatical installation (e.g. kickstart)
is allowed to start.

At present this can only be accomplished with ctrl-c breaking the playbook and resetting the system in the vmware_gui.

The state "poweredon|powerdoff|present|..." is used in a strange way:
Setting it to "powerdoff" while CREATING a new vmware guest is simply ignored. But the same state: poweredoff an a ALREADY CREATED vmware_guest really switch off the guest NOW.

So this state can only be used with a existing guest, it is of no use (or even contrary to my whish)
to a to be created vmware guest:

Proposal: Add a flag (e.g. creation_state: [powerdon|poweredoff] default('powerdon')) to the CREATING of a vmware_guest, how the state after creation
of a new guest should be. IGNORE this flag on a INSTALLED guest.

And/Or make creation of guest possible with two or more different datastores.
Also please explain, which flags only work while creating, which work while updating.
(Updating a "created at date_time string" in annotation makes no sense in subsequent playbook runs)

**Comment by ngp-star** [comment]
hi @harvey637 

can you please elaborate on the following proposal what you suggested.

Proposal: Add a flag (e.g. creation_state: [powerdon|poweredoff] default('powerdon')) to the CREATING of a vmware_guest, how the state after creation
of a new guest should be. IGNORE this flag on a INSTALLED guest.

**Comment by pgbidkar** [comment]
+label needs_info

**Comment by harvey637** [comment]
hi @ ngp-star
when CREATING a vmware_guest this new guest is normally started after creation.
Mostly ok so, BUT:
If you need to create a vmware_guest that ist NOT powered on it is not possible in ansible. You have to go to the vmware_host to immediatly stop the system. Than you are able to modify the newly created vmware_guest to modify settings, that are not handled (or not correctly handled) in ansible.

For me this was the problem with creating vmware_guests with more than one disk, where the location of the disk was cruical to the system, e.g. disk 3 needs to be on an SSD, where all other disks are on rotational disks. Possibly this problem was remedied by now, havent checked lately.
 So to elaborate on this:
CREATION of new vmware_guest finishes in a desired power_state, default is the old behavior: poweredON. With this flag a newly created vmware_guest could be left in the state poweredOFF.
This flag ONLY affects the CREATION of a new vmware_guest, it has no meaning an an allready created vmware_guest.
Got my idea?
cu
Harvey

**Comment by pgbidkar** [comment]
Hi @harvey637 , Thanks for the details. if I understand it correctly here vmware_guest means virtual machine. You are trying to create virtual machine in poweredoff state. 

vmware_guest module has flag to create VM in powered off state.
Please try task like this 
```
- name: Set powerstate of a virtual machine to poweroff by using UUID
  vmware_guest:
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    validate_certs: no
    uuid: "{{ vm_uuid }}"
    state: poweredoff
  delegate_to: localhost
```

or in your playbook set the state parameter to poweredoff.

Please let me know if more details required
Thanks

**Comment by pgbidkar** [comment]
+label needs_info

**Comment by lparkes** [comment]
This is what the default value of ``state`` (i.e. ``present``) already does. I use this all the time to create new guests in a powered off state so that I can configure various other things needed to kickstart them and then I power on the guest. I'm running Ansible 2.8.x. 

Here is a snippet from main.yaml of the role I use.

```yaml
- name: Create a new VM
  vmware_guest:
    datacenter: "{{ datacentre }}"
    folder: "{{ folder }}"
    cluster: "{{ cluster }}"
    datastore: "{{ datastore }}"
    disk: "{{ disks }}"
    guest_id: "{{ rhel }}_64Guest"
    hardware:
      memory_mb: "{{ ram | int * 1024 }}"
      num_cpus: "{{ cpus }}"
      hotadd_cpu: yes
      hotadd_memory: yes
      memory_reservation_lock: yes
    name: "{{ inventory_hostname }}"
    annotation: "{{ notes }}"
 
    hostname: "{{ vcenter }}"
    username: "{{ vmware_user }}"
    password: "{{ vmware_password }}"
 
    resource_pool: "{{ resource_pool|default(omit) }}"
 
    networks:
      - name: "{{ network }}"
        device_type: vmxnet3
 
  register: deploy
  delegate_to: localhost
  become: no

- name: Configure TFTP installation of new VM
  template:
    src: pxe.conf.{{ rhel }}.j2
    dest: "{{ tftp_dir }}/01-{{ deploy.instance.hw_eth0.macaddress_dash }}"
    owner: root
    group: root
    mode: 0644
  delegate_to:  "{{ tftp_server }}"
  become: yes
  when: deploy.instance.hw_power_status == "poweredOff"
 
- name: Boot the VM
  local_action:
    module: vmware_guest
    datacenter: "{{ datacentre }}"
    use_instance_uuid: yes
    uuid: "{{ deploy.instance.instance_uuid }}"
    state: poweredon
    wait_for_ip_address: yes
    state_change_timeout: 1200
 
    hostname: "{{ vcenter }}"
    username: "{{ vmware_user }}"
    password: "{{ vmware_password }}"
 
  register: boot
  become: no
 
- name: Unconfigure TFTP installation of new VM
  file:
    dest: "{{ tftp_dir }}/01-{{ deploy.instance.hw_eth0.macaddress_dash }}"
    state: absent
  delegate_to: "{{ tftp_server }}"
  become: yes
```

**Comment by Akasurde** [comment]
@harvey637 Do you agree with @pgbidkar and @lparkes opinions ? If yes, let me know so that we close this issue. 

needs_info

**Comment by ansibot** [comment]
@harvey637 This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@harvey637 You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
